### ISSUE TITLE
Strange inferred types with loops

### ISSUE THREAD

**Issue by Timmmm** [microsoft/pyright#3074]
This is via VSCode with Pylance v2022.2.3 (latest version as of today).

Take a look at these two functions:

```python
def fn1(foo: list[str | None]) -> bool:

    i = 0
    x: Optional[int] = None

    while i < 5:
        foo_part = foo[i]

        if foo_part is None:
            i += 1
            x = i
            continue

        if x is None:
            return False
        i = x

    return True


def fn2(foo: list[str]) -> bool:

    i = 0
    x: Optional[int] = None

    while i < 5:
        foo_part = foo[i]

        if x is None:
            return False
        i = x

    return True
```

As far as I can tell by inspection, all their types are fine and it should be possible to infer all of the types. But the first function gives some warnings:

![image](https://user-images.githubusercontent.com/376842/154688527-626af41b-9cab-4307-8d19-5e255d22e6a7.png)

And the second one gives errors:

![image](https://user-images.githubusercontent.com/376842/154688575-f5d76a02-2cf2-4d09-9258-bce15e22bc55.png)

Sorry if I'm being stupid and missing something obvious!

**Comment by Timmmm** [comment]
Here's a probably-related tricky one:

```python
def fn3(a: list[str | None]) -> bool:
    a_i = 0
    b_i: Optional[int] = None

    while a_i < 5:
        a_part = a[a_i]

        if a_part is None:
            a_i += 1
            b_i = a_i
            continue

        if a_part == "":
            a_i += 1
        else:
            if b_i is None:
                return False
            a_i = b_i

    x = a_i == len(a)
    return x
```

It gives a warning that the type of `x` is unknown:

![image](https://user-images.githubusercontent.com/376842/154689445-63a8a8dd-e537-4aef-a425-f550855534c4.png)

Despite the fact that it correcly inferred `a_i` is an `int`:

![image](https://user-images.githubusercontent.com/376842/154689519-70fa2d60-0bcd-45b2-9858-f83644709ceb.png)

But if you change `a_i = 0` to `a_i: int = 0` then the warning goes away!

There's various other very strange behaviours, e.g. if you remove the first `a_i += 1` line then you get type errors.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by Timmmm** [comment]
Amazing, thanks!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.223, which I just published. It will also be included in the next release of pylance.
